homeassistant/components/esphome/alarm_control_panel.py

```diff
@@ -0,0 +1,135 @@
+"""Support for ESPHome Alarm Control Panel."""
+from __future__ import annotations
+
+from aioesphomeapi import (
+    AlarmControlPanelCommand,
+    AlarmControlPanelEntityState,
+    AlarmControlPanelInfo,
+    AlarmControlPanelState,
+)
+
+from homeassistant.components.alarm_control_panel import (
+    AlarmControlPanelEntity,
+    AlarmControlPanelEntityFeature,
+    CodeFormat,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    STATE_ALARM_ARMED_AWAY,
+    STATE_ALARM_ARMED_CUSTOM_BYPASS,
+    STATE_ALARM_ARMED_HOME,
+    STATE_ALARM_ARMED_NIGHT,
+    STATE_ALARM_ARMED_VACATION,
+    STATE_ALARM_ARMING,
+    STATE_ALARM_DISARMED,
+    STATE_ALARM_DISARMING,
+    STATE_ALARM_PENDING,
+    STATE_ALARM_TRIGGERED,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import EsphomeEntity, EsphomeEnumMapper, platform_async_setup_entry
+
+_ESPHOME_ACP_STATE_TO_HASS_STATE: EsphomeEnumMapper[
+    AlarmControlPanelState, str
+] = EsphomeEnumMapper(
+    {
+        AlarmControlPanelState.DISARMED: STATE_ALARM_DISARMED,
+        AlarmControlPanelState.ARMED_HOME: STATE_ALARM_ARMED_HOME,
+        AlarmControlPanelState.ARMED_AWAY: STATE_ALARM_ARMED_AWAY,
+        AlarmControlPanelState.ARMED_NIGHT: STATE_ALARM_ARMED_NIGHT,
+        AlarmControlPanelState.ARMED_VACATION: STATE_ALARM_ARMED_VACATION,
+        AlarmControlPanelState.ARMED_CUSTOM_BYPASS: STATE_ALARM_ARMED_CUSTOM_BYPASS,
+        AlarmControlPanelState.PENDING: STATE_ALARM_PENDING,
+        AlarmControlPanelState.ARMING: STATE_ALARM_ARMING,
+        AlarmControlPanelState.DISARMING: STATE_ALARM_DISARMING,
+        AlarmControlPanelState.TRIGGERED: STATE_ALARM_TRIGGERED,
+    }
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up ESPHome switches based on a config entry."""
+    await platform_async_setup_entry(
+        hass,
+        entry,
+        async_add_entities,
+        component_key="alarm_control_panel",
+        info_type=AlarmControlPanelInfo,
+        entity_type=EsphomeAlarmControlPanel,
+        state_type=AlarmControlPanelEntityState,
+    )
+
+
+class EsphomeAlarmControlPanel(
+    EsphomeEntity[AlarmControlPanelInfo, AlarmControlPanelEntityState],
+    AlarmControlPanelEntity,
+):
+    """An Alarm Control Panel implementation for ESPHome."""
+
+    @property
+    def state(self) -> str | None:
+        """Return the state of the device."""
+        return _ESPHOME_ACP_STATE_TO_HASS_STATE.from_esphome(self._state.state)
+
+    @property
+    def supported_features(self) -> AlarmControlPanelEntityFeature:
+        """Return the list of supported features."""
+        return AlarmControlPanelEntityFeature(self._static_info.supported_features)
```

Comment:
We need to translate esphome supported features to hass supported features as esphome has not guarantee that the supported features will always match hass.

fan and cover have an example